sdk/clientcore/annotation-processor/src/main/java/io/clientcore/annotation/processor/utils/ResponseHandler.java

```diff
@@ -40,50 +43,98 @@ public final class ResponseHandler {
+        if (TypeUtil.isTypeOrSubTypeOf(entityType, InputStream.class)
+            || TypeUtil.isTypeOrSubTypeOf(entityType, BinaryData.class)) {
+            return false;
+        }
+
+        return entityType == Void.TYPE
+            || entityType == Void.class
+            || (method.getHttpMethod() == HttpMethod.HEAD && isBooleanType(entityType))
+            || TypeUtil.isTypeOrSubTypeOf(entityType, byte[].class);
     }
 
-    private static void handleRequestReturn(BlockStmt body, TypeMirror returnType, HttpRequestContext method,
-        boolean serializationFormatSet) {
-        java.lang.reflect.Type entityType = TypeConverter.getEntityType(returnType);
+    private static void validateResponseStatus(BlockStmt body, HttpRequestContext method,
+        boolean usingTryWithResources) {
+        body.addStatement(StaticJavaParser.parseStatement("int responseCode = networkResponse.getStatusCode();"));
+        String expectedResponseCheck;
+        if (method.getExpectedStatusCodes().size() == 1) {
+            expectedResponseCheck = "responseCode == " + method.getExpectedStatusCodes().get(0) + ";";
+        } else {
+            String statusCodes = method.getExpectedStatusCodes()
+                .stream()
+                .map(code -> "responseCode == " + code)
+                .collect(Collectors.joining(" || "));
+            expectedResponseCheck = "(" + statusCodes + ");";
+        }
+        body.addStatement(StaticJavaParser.parseStatement("boolean expectedResponse = " + expectedResponseCheck));
+
+        body.tryAddImportToParentCompilationUnit(HttpResponseException.class);
+        BlockStmt ifBlock = new BlockStmt();
+        ifBlock.addStatement(
+            StaticJavaParser.parseStatement("String errorMessage = networkResponse.getValue().toString();"));
+        if (!usingTryWithResources) {
+            closeResponse(ifBlock);
+        }
+        ifBlock.addStatement(
+            StaticJavaParser.parseStatement("throw new HttpResponseException(errorMessage, networkResponse, null);"));
+        IfStmt ifStmt = new IfStmt()
+            .setCondition(new UnaryExpr(new NameExpr("expectedResponse"), UnaryExpr.Operator.LOGICAL_COMPLEMENT))
+            .setThenStmt(ifBlock);
+        body.addStatement(ifStmt);
+    }
+
+    private static void handleRequestReturn(BlockStmt body, TypeMirror returnType, java.lang.reflect.Type entityType,
+        HttpRequestContext method, boolean serializationFormatSet) {
         boolean returnIsResponse = TypeConverter.isResponseType(returnType);
 
         // TODO (alzimmer): Base64Uri needs to be handled. Determine how this will show up in code generation and then
         //  add support for it.
-        if (entityType == Void.TYPE || entityType == Void.class) {
-            // Return is void, Void, or Response<Void>, close the network response and return null as the value.
+        if (returnType.getKind() == TypeKind.VOID) {
```

Comment:
Does this mean we only close the response explicitly if the method returns `void`? Would we keep it open for `Void` or `Response<Void>`? Why?

Reply:
Added comments to explain this further.